MSC2666: Get rooms in common with another user

Half-Shot opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-doc/blob/hs/shared-rooms/proposals/2666-get-rooms-in-common.md)

Author: @Half-Shot, @ShadowJonathan 

Implementations:
- Synapse: https://github.com/matrix-org/synapse/pull/15621 + https://github.com/element-hq/synapse/pull/19279
- Ruma: https://github.com/ruma/ruma/pull/1547
- JS SDK: https://github.com/matrix-org/matrix-js-sdk/pull/3381

<details><summary>Click to reveal Legacy Implementations</summary>

- matrix-react-sdk PR: matrix-org/matrix-react-sdk#4897
- matrix-js-sdk PR: matrix-org/matrix-js-sdk#1417
- Synapse (Went through a few iterations)
  - https://github.com/matrix-org/synapse/pull/7785
  - https://github.com/matrix-org/synapse/pull/12036
  - https://github.com/matrix-org/synapse/pull/12038
  - https://github.com/matrix-org/synapse/pull/12836
</details>

---

# Spec PR note

The endpoint path was clarified in https://github.com/matrix-org/matrix-spec-proposals/commit/9a49a4779d2f467ce853df296c16b65b4ddc0b8a

---

SCT stuff:

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#issuecomment-3614930041)

[MSC checklist](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#issuecomment-3614713505)

Half-Shot committed:
MSC 2664: Get rooms in common with another user

Half-Shot committed:
correct MSC number

Sorunome commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,71 @@
+```
+
+```json
+[
````
Shouldn't the reply be

```json
{
  "rooms": [
    "!blah",
    "!blubb",
  ]
}
```
As so far *no* endpoint in the spec returns an array, only objects?

Half-Shot replied:
Yeah, agreed. I was lazy here.

Sorunome commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,71 @@
+Homeservers should implement a new endpoint `/users/{user_id}/shared_rooms/{other_user_id}` which will take
+the current users mxid and the other users mxid, which the user is trying to search for.
+
+The response format will be an array containing all rooms where both the `user_id` and `other_user_id` have
```
It should be required that one of the user IDs is your own user ID, thus also requiring authentication via access token

tulir replied:
It would probably make most sense to require `user_id` to be your own user ID

Half-Shot replied:
I was riffing off what we do for account_data. My thought process was as follows:

- Seems a shame to add a new root `/shared_rooms` path, I wonder if I can use `/rooms`.
- Hm, no. `/rooms` is for fetching data about a single room. How about `/users`
- Oh, but all `/users` calls expect the owners user_id.

So in the end I could either break the `/users` convention, write a new path, or use the existing convention. I went with C, but I am happy to hear arguments otherwise.

Half-Shot replied:
> It would probably make most sense to require `user_id` to be your own user ID

That's what I am doing in the code, perhaps I need to spell it out :P

timokoesters replied:
In conduit I fully ignore the user id in the path and always use the user id from the access token. Is that okay?

Half-Shot replied:
In practise that's fine. It's a bit of a spec wart that we have the user_id in the path, because the spec does not allow for it to differ from the access_token.

timokoesters commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,71 @@
+the current users mxid and the other users mxid, which the user is trying to search for.
+
+The response format will be an array containing all rooms where both the `user_id` and `other_user_id` have
+a membership of type `join`. 
```
Wouldn't it also be useful to return rooms which you both participated in, in the past, but the other user was banned/kicked. So instead of returning a list of rooms you both joined, the endpoint could return multiple lists, one for each membership state

Half-Shot replied:
In theory we could stick a ?membership=[types] on the end of this endpoint, which is a comma separated list of allowed membership states for the remote user. I guess we could even do it for the current user too, however I'm unsure how useful this is.

Sorunome replied:
This unfortunately leaks information (and thus soru is strongly in favour *not* to do it):

Suppose alice and bob shared a room at once point, but alice left. Via that endpoint she can then keep querying the room if bob is still in there.

timokoesters replied:
@Sorunome I meant room where alice is joined, but bob has a different membership state. In the UI it would show all rooms alice joined and bob has *any* membership event in, with some text saying which one it is.

Sorunome replied:
Aaaah, that would be ok, as in, that wouldn't leak information, thanks for clarification!

timokoesters replied:
Another idea is to also show the power level, sothe Ui can say that a user is moderator, admin or muted in some room

Sorunome replied:
That would also require fetching power level events for a given room, not only search `m.room.member`, and thus would (greatly?) increase complexity on the server side for little gain?

timokoesters replied:
I don't think getting two events of the current state is expensive, but maybe in synapse's architecture it is. Idk

Half-Shot replied:
I think it's important to keep this MSC scoped in order to improve it's chances of ~~not getting bikeshedded~~ keeping the implementation easy for *all* homeservers.

I'm going to consciously prioritise the use case of showing a simple list of rooms that you are both in, and clients can populate their UI with any state they have for the room.

timokoesters replied:
Perhaps you can rename the `rooms` field to `join` then, so it's easy to add to it in the future?

Half-Shot replied:
Yep, I think that's a fair compromise.

Half-Shot replied:
I've gone with "joined" to line up with things like https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-rooms-roomid-joined-members, but otherwise it's been included.

Half-Shot committed:
Try to clarify Proposal, update response format

Half-Shot committed:
Update MSC number in prefix

Half-Shot committed:
Wording tidyup

Half-Shot committed:
more tidyup

Half-Shot committed:
Clarify empty responses

uhoreg commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,72 @@
+rooms you have in common before you accept an invite.
+
+While this information can be determined if the user has full access to member state for all rooms,
+modern clients often implement "lazy-loadeding", so they often only have state for a subset of all rooms
```
```suggestion
modern clients often implement "lazy-loading", so they often only have a subset of state for the rooms
```

uhoreg commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,72 @@
+## Proposal
+
+Homeservers should implement a new endpoint `/users/{current_user}/shared_rooms/{other_user_id}` which will take
+the authenticated users MxID and the user that is being searched for.
```
```suggestion
the authenticated user's MxID and the user that is being searched for.
```

Half-Shot committed:
uhoreg fixes my spelling

Co-authored-by: Hubert Chathi <hubert@uhoreg.ca>

timokoesters commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,72 @@
+the authenticated user's MxID and the user that is being searched for.
+
+The response format will be an array containing all rooms where both the `current_user` and `other_user_id` have
+a membership of type `join`. If either user is not joined to any rooms, or the `other_user_id` does not exist, an
```
What happens if current_user == other_user ?

Half-Shot replied:
Good point. In my opinion it's an invalid request, as you can just called /joinedRooms for the same result except that path can optimise for the current user.

auscompgeek replied:
So something like this?

Servers may reject the request, with [some appropriate error code], if `other_user_id` is the same as `current_user`.  Clients should use /joinedRooms to determine what rooms the current user has joined instead.

Half-Shot replied:
It's a tossup between consistency and being too strict. I am minded to be lean on a strong requirement to reject, because doing this has no purpose since another endpoint exists.

Half-Shot committed:
add same user case

Half-Shot committed:
Update to use /uk.half-shot.msc2666/

Half-Shot commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,73 @@
+reject with HTTP 400.
+
+```
+GET _matrix/client/unstable/uk.half-shot.msc2666/users/@alice:example.com/shared_rooms/@bob:example.com
````
@anoadragon453 asks if we should continue to use the /users/{user_id}/.../{other_user_id}/ format given we don't allow you to do it on behalf of other users. This was probably intended as a way for appservices to puppet their accounts a long time ago before as_user became the way to do things like this

Half-Shot replied:
IMO, I'm fine with dropping this OR we could decide that homeservers can decide whether to allow other user_ids so that admins can spy on their users?

anoadragon453 replied:
I was thinking about this functionality, but I can't really visualize how this would be useful in client UI.

I could see something that would allow you to see what rooms an arbitrary number of users share though? That's probably out of scope here though.

Admins can also usually just go and have a poke in the DB as well.

Half-Shot replied:
Yeah, I think at best this would be convenience but frankly it means every client now needs to paste in their user_id in order to call the api. That sucks. Let's drop it.

timokoesters replied:
I'm for dropping as well. Maybe even replace existing endpoints with non-userid paths (or support both for a while)

uhoreg commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,73 @@
+
+## Proposal
+
+Homeservers should implement a new endpoint `/users/{current_user}/shared_rooms/{other_user_id}` which will take
```
Why `/users/` rather than `/user/` as other existing endpoints use?

tulir replied:
Seems to be a typo in the MSC, it's implemented as `/user/`

Half-Shot replied:
fixed

tulir commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,73 @@
+reject with HTTP 400.
+
+```
+GET _matrix/client/unstable/uk.half-shot.msc2666/users/@alice:example.com/shared_rooms/@bob:example.com
````
Isn't the MSC supposed to use the actual endpoint and not vendor-prefixed ones?

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,73 @@
+
+## Alternatives
+
+A client can already read all membership for all rooms, and thus determine which of those rooms contains
+a "join" membership for the given user_id. However, this method is computationally expensive on the homeserver
+and the client. Furthermore, it would increase total network traffic (which is important for low bandwith / mobile clients)
+as well as include lots of extranious information.
```
This does not feel like a compelling reason to add an endpoint to the spec. Low bandwidth and mobile clients should be using more restrictive filters to limit their exposure to information, and if a client isn't using a local store for checking this then they've really messed up how this feature could work. Clients which would want this feature would have already cached state locally.

tulir replied:
I don't really get your point. Do you mean there should be some non-standalone-endpoint way to get the information without loading all room state?

The only way for this to work without any API changes is if clients either stop lazy loading or request the member lists for *all* rooms at the same time. What's wrong with wanting this feature and lazy loading at the same time?

turt2live replied:
Lazy-loading is a much stronger justification than the words here.

Half-Shot replied:
Lazy-loading is mentioned [here](https://github.com/matrix-org/matrix-doc/pull/2666/files/db99583ce577582f6bfe4b28d6ecfa7b1bf54590#diff-366ea8b6ee99e7bf4f9501ae6e9ac376R13-R15) and it was meant to be taken in mind when reading the rest of the MSC, but I'll make a point of including it in this section too.

turt2live reviewed (changes requested):
Overall this doesn't feel like it needs to be an endpoint.

Half-Shot committed:
Fixes

Half-Shot committed:
Merge branch 'hs/shared-rooms' of github.com:matrix-org/matrix-doc into hs/shared-rooms

Half-Shot committed:
remove referneces to user_id

Half-Shot committed:
another reference to the auth user id

Half-Shot committed:
consistent newlines

Half-Shot commented:
Synapse and matrix-js-sdk have been updated to match the latest version of the spec.

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,76 @@
+  ]
+}
+```
+
````
Could you specify some expected error codes here please? Such as the code when trying to request shared rooms with yourself.

Half-Shot replied:
I think there is only one error situation likely to occur, which is the one you mentioned. Users who do not exist will return an empty list, as will invalid user_ids.

anoadragon453 replied:
I probably would've said `M_INVALID_PARAM`, but I don't think it matters much.

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,76 @@
+## Unstable prefix
+
+The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/shared_rooms/{user_id}`.  
+The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666`.  
```
Just checking that you're stating that `uk.half-shot.msc2666` has to be present, but could either be `true` or `false` to indicate that this homeserver supports this feature.

Half-Shot replied:
Have spelt it out, but yes.

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,76 @@
+an empty array should be returned.
+
+```
+GET _matrix/client/r0/user/shared_rooms/@bob:example.com
````
```suggestion
GET /_matrix/client/r0/user/shared_rooms/%40bob%3Aexample.com
```

Half-Shot committed:
Add errcode

Half-Shot committed:
Update proposals/2666-get-rooms-in-common.md

Co-authored-by: Andrew Morgan <1342360+anoadragon453@users.noreply.github.com>

timokoesters commented:
One can implement this very efficiently if one already stores a sorted list of joined roomids for both users, because then you can go through both lists at the same time resulting in linear time complexity

anoadragon453 commented:
> One can implement this very efficiently if one already stores a sorted list of joined roomids for both users, because then you can go through both lists at the same time resulting in linear time complexity

Nice tip! One to go in the [homeserver implementation book](https://spec.matrix.org/books/server/) (once it's fleshed out a bit...).

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,78 @@
+}
+```
+
+The response error for trying to get shared rooms with yourself will be an HTTP code 400, with `M_FORBIDDEN` as the `errcode`.
````
Should this be an HTTP `403` since it's `M_FORBIDDEN`?

Half-Shot replied:
Doesn't have to be. [403](https://httpstatuses.com/403) "The server understood the request but refuses to authorize it." would imply an authentication failure which would leave clients confused because they are using correct authentication information.

[400](https://httpstatuses.com/400) "The server cannot or will not process the request due to something that is perceived to be a client error" which would be correct here, since the request cannot be processed due to incorrect parameters. The request is still FORBIDDEN, but not due to authentication. This is the closest approximation I could find in the spec. `M_INVALID_PARAM` would have worked, but the spec implies that this is due to incorrect typing which isn't the case here.

I think this is fine.

ptman replied:
I've used 422 Unprocessable Entity in some of my own projects.

Half-Shot replied:
422 would actually make the most sense, but I couldn't find any usage of it in Synapse/TheSpec so I assumed it was unpopular (as 410+ error codes often are..)

Half-Shot commented:
Heads up that the Synapse PR has landed as an unstable feature now.

ShadowJonathan commented:
Blocked on https://github.com/matrix-org/synapse/issues/10339

kevincox commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,78 @@
+
+## Proposal
+
+Homeservers should implement a new endpoint `/user/shared_rooms/{user_id}` which will take
```
Nit: `shared_rooms` could be taken to mean either "rooms I have shared with this user" or "rooms that I share with this user". It is probably better to call this `common_rooms`. (Although that may be confused for common as in popular, but that doesn't make sense when passing another user ID).

ShadowJonathan replied:
I kinda agree, we could probably rename it to "mutual_rooms", just like discord has "mutual friends/servers".

kevincox replied:
I think that name is quite clear.

Half-Shot replied:
Mutual sounds good to me, and probably the most clearest one.

kevincox commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,78 @@
+and /search may be complimentary.
+
+
+## Alternatives
```
Does it make sense to extend the search API with member constraints?

This would be more generic and be able to specify multiple usersto include? This may be helpful for finding old rooms (I remember that I was walking with Alice and Bob about this but can't find the room).

There is a bit of a mismatch because right now because IIUC the search API always returns messages however I can also imagine this filter being useful for message searches. (Ex: messages containing "pizza" in a room that contains Alice and Bob.) Extending the Search API with a `result_type: room` option and adding support for a `member` constraint would add both of these features and provide an obvious way to add more constraints on "Find common rooms" in the future.

tulir replied:
The `/search` API seems to already have query and result categories (`room_events` being the only currently defined one). It also already supports sender constraints, but that doesn't directly help with room search.

kevincox replied:
That sounds good. So we could add the `rooms` category and add `member` constrains. I agree that `sender` is different than `member`.

Although this does raise fallback questions. If we say that `member` can be used for `room_events` is it required? Or can the server ignore it. But is a server allowed to ignore it for `rooms`? If it does ignore it it makes it useless for the use case described by this MSC.

ShadowJonathan replied:
While i think these are good ideas, they kinda fall out of scope of this MSC, the search api needs better tuning in general, and i don't think that bolting something like this onto it is a good idea for now. We should probably revisit this bit when we overhaul the search API in a more broader sense.

Half-Shot replied:
+1 I'd rather not encumber the search API with more things when it's already not fit for purpose, and would rather we wrote a MSC specifically to alter how search works separately.

anoadragon453 replied:
Off the back of the above conversation, are there any use cases for allowing clients to hand multiple user IDs to this API, that act as an AND condition to search mutual rooms with?

All I can think of is the use case that @kevincox mentioned above, where you're trying to find a room given some known participants - which actually sounds like a pretty cool feature. We would need to reshape this API to use query parameters instead of path parameters or similar though. Do others think that may be worthwhile, or can think of other use cases?

It's also perfectly valid to just assume that specific use case will be handled by a more featureful room search API.

ShadowJonathan replied:
I think that at this point, it's too much of a diversion from what has already settled in over the course of this MSC's development; personally, i'd like to see that in a new endpoint, or in that room search api you mentioned.

(FWIW, the above is talking about a message search API as well, i'm mentioning this to make sure you're not confusing the above two.)

anoadragon453 replied:
I appreciate that this MSC has been open for a while, but I'm not sure if it's _actually_ near-complete. We've yet to reach proposed FCP or have any implementation in use other than [a trivially mutable function](https://github.com/matrix-org/matrix-js-sdk/pull/1417/files#diff-627fe08d8bba6a4fa76e7c5ed36ef6f16d126733e65e19236b391de0d34f1fe0R4852-R4864) written for it. I don't think potentially major changes to the MSC at this point are off the table?

I know that extending this MSC's lifecycle isn't ideal, but I do want to make sure we have a maintainable API.

My current thoughts are:

* If we *do* end up wanting this functionality (**which is not yet clear!**) then we'd be painting ourselves into a corner with the current shape.
* The convention of putting user IDs in path parameters was originally intended to allow AS' to insert puppeted Matrix User ID. But that ended up being scrapped for [the `user_id` query parameter](https://spec.matrix.org/v1.3/application-service-api/#identity-assertion). As a result, ideally we'd remove about half of the instances of user IDs in path parameters; thus it may not be the best precedence to adhere to.

> (FWIW, the above is talking about a message search API as well, i'm mentioning this to make sure you're not confusing the above two.)

Ah indeed it is talking about message search. My comment probably should have been its own thread.

ShadowJonathan replied:
We discussed in DMs what the best way forward for this MSC would be, and we came to the conclusions;
- To swap the path parameter to a query parameter.
- To have such an endpoint implicitly include the self-user in the search.
- A possible moderation use-case for this can best be addressed in an admin api.
- If, in the future, a use case comes along where more-than-one users have to be queried, then this endpoint can be easily edited by another MSC, and clients will have to version-check if the server supports multiple users with the query part.

ShadowJonathan replied:
/res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

ShadowJonathan committed:
typo fix and shared_rooms -> mutual_rooms (#3631)

ShadowJonathan commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,78 @@
+an empty array should be returned.
+
+```
+GET /_matrix/client/r0/user/mutual_rooms/%40bob%3Aexample.com
````
There should probably be a way to refuse service to this endpoint with no data, see https://github.com/matrix-org/synapse/pull/12038#discussion_r819599640

I suggest responding with a 404 as a known general failure condition, in addition to "user not found".

anoadragon453 replied:
```suggestion
GET /_matrix/client/v1/user/mutual_rooms/%40bob%3Aexample.com
```

To follow the new per-endpoint versioning scheme.

ShadowJonathan commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,78 @@
+# MSC 2666: Get rooms in common with another user
```
@ spec scribe;

I've added an implementation which updates the JS SDK code to use the new naming; https://github.com/matrix-org/matrix-js-sdk/pull/2271

ShadowJonathan commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,78 @@
+## Unstable prefix
+
+The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/mutual_rooms/{user_id}`.
+The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666`.
```
```suggestion
The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666.mutual_rooms`.
```

See [this implementation review comment](https://github.com/matrix-org/matrix-js-sdk/pull/2271#discussion_r841389852), considering we now have two endpoints for one unstable MSC, we should differentiate the two as well.

Half-Shot replied:
Ah, I see. Yes, this seems valid if we've changed the endpoint between versions.

Half-Shot committed:
Apply suggestions from code review

Co-authored-by: Jonathan de Jong <jonathandejong02@gmail.com>

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,78 @@
+The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/mutual_rooms/{user_id}`.
+The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666.mutual_rooms`.
+If the value is false or the key is not present, clients MUST assume the feature is not available.
+Once the MSC has been merged, clients should use `/_matrix/client/r0/user/mutual_rooms/{user_id}`
```
```suggestion
Once the MSC has been merged, clients should use `/_matrix/client/v1/user/mutual_rooms/{user_id}`
```

anoadragon453 reviewed (commented):
Other than working through the minor case in https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r830500133 and the below nits, I think this MSC is ready for FCP.

What say you @Half-Shot?

ShadowJonathan committed:
@MSC2666: Add "may return 400" qualifier (#3770)

Resolves [this comment](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r830500133)

Rationale for `M_UNKNOWN`: There's not a "semantically more close" variant of that error, and i'm hesitant to add a new error.

cc @Half-Shot

Half-Shot committed:
Reject r0, embrace v1

Co-authored-by: Andrew Morgan <1342360+anoadragon453@users.noreply.github.com>

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,81 @@
+a membership of type `join`. If the `user_id` does not exist, or does not share any rooms with the authenticated user,
+an empty array should be returned.
+
+Additionally, a server may respond with a `400` `M_UNKNOWN` to this endpoint for any reason, 
+signalling its unwillingess (or inability) to satisfy the request.
```
this is a very uncommon thing to do in the spec. All endpoints are currently required to be implemented (unless bundled into a module the server doesn't care about), so this endpoint should never fall into "unwillingness" or "inability". 

If we do want to do this, please provide justification in the MSC.

ShadowJonathan replied:
I'm conflicted on this requirement myself also, I'm starting to think that the arbitrary requirement from Synapse - where this is implemented via the optionally-ran user directory - would not be enough to restrict this MSC to have to include this.

I will take another look at the Synapse implementation, see if it can be implemented in a way that makes this section redundant.

turt2live replied:
Conveniently, the Synapse implementation is what drove this thread being opened. Implementation-specific restrictions are not supposed to be carried into the protocol (but cases where many/all implementations have issues would be considered).

For this case, the Synapse side is kinda stuck at "figure it out" rather than having the spec concede.

ShadowJonathan replied:
https://github.com/matrix-org/synapse/pull/12836 pushed a change to synapse, I'll keep this open until that has been resolved.

ShadowJonathan replied:
This has been resolved on both fronts now.

ShadowJonathan replied:
https://github.com/matrix-org/matrix-spec-proposals/pull/3822 has also been made to address the proposal-text-side of things.

ShadowJonathan replied:
/res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

Half-Shot commented:
Hi there, I'm delegating this PR to @ShadowJonathan as I lack time to complete it.

ShadowJonathan committed:
Remove M_UNKNOWN response (#3822)

ShadowJonathan commented:
I hereby sign all my comments on this branch, preceding and following, with;

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

turt2live commented:
The author/shepherd, @ShadowJonathan, believes this MSC is ready for FCP.

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,78 @@
+The implementation MUST use `/_matrix/client/unstable/uk.half-shot.msc2666/user/mutual_rooms/{user_id}`.
+The /versions endpoint MUST include a new key in `unstable_features` with the name `uk.half-shot.msc2666.mutual_rooms`.
+If the value is false or the key is not present, clients MUST assume the feature is not available.
+Once the MSC has been merged, clients should use `/_matrix/client/v1/user/mutual_rooms/{user_id}`
```
```suggestion
Once the MSC has been merged, and the homeserver has advertised support for the Matrix version that this endpoint is included in, clients should use `/_matrix/client/v1/user/mutual_rooms/{user_id}`
```

ShadowJonathan replied:
👍

@Half-Shot Could you add this?

ShadowJonathan replied:
/res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

anoadragon453 committed:
Update an old link to the spec; fix endpoint text

anoadragon453 committed:
provide a link to lazy-loading of room members spec

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,79 @@
+Homeservers should implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take
+the MxID of the user who is being searched for.
+
+The response format will be an array containing all rooms where both the authenticated user and `user_id` have
+a membership of type `join`. If the `user_id` does not exist, or does not share any rooms with the authenticated user,
+an empty array should be returned.
```
With room types now merged to the spec, should we add a query parameter to constrain the results by type of room? I could see this being useful for:

1. Being able to show a separate list of "Mutual Spaces" and "Mutual rooms".
1. Hiding rooms from the results that shouldn't be shown in clients (i.e. profile rooms).

I also recognise that the client could probably do local filtering, but I'm not sure how different that would be to implement.

ShadowJonathan replied:
Personally, I'd like this to be implemented in the Room Search API you alluded to [here](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r935730582), in a seperate MSC, rather than tacking it onto this one.

I think going forward, local filtering and matching could better address this usecase than not, I think going forward with this MSC as-is is a better option than to wait a longer while for a more fruitful Room Search API.

anoadragon453 reviewed (commented):
From the PR description:

> matrix-react-sdk PR: https://github.com/matrix-org/matrix-react-sdk/pull/4897

It looks like this PR was never merged. I think it's useful to leave it for historical purposes, but could you indicate that this PR was closed?

Otherwise this is looking positive overall, and I think just mainly need some more client implementation work to push it over the edge (the js-sdk PR is a good start, but I'd be interested in seeing how a client implements UI for this, especially with respect to different room types).

ShadowJonathan commented:
> the js-sdk PR is a good start, but I'd be interested in seeing how a client implements UI for this, especially with respect to different room types

Element's progression is tracked [here](https://github.com/vector-im/element-meta/issues/371) and [here](https://github.com/vector-im/element-web/issues/14339), seeing as they were the original driver of this MSC, i requested the design team if they could pick this up in their new team structure.

Seeing as it is available in js-sdk, cinny.in could implement this themselves (tracked in https://github.com/cinnyapp/cinny/issues/710)

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,79 @@
+
+## Proposal
+
+Homeservers should implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take
```
```suggestion
Homeservers implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take
```

"should" implies it's optional when it's not :P

ShadowJonathan replied:
/res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,79 @@
+
+## Proposal
+
+Homeservers should implement a new endpoint `/_matrix/client/v1/user/mutual_rooms/{user_id}` which will take
```
can it be rate limited?

ShadowJonathan replied:
Yes, that should be added. 👍

ShadowJonathan replied:
/res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,79 @@
+
+## Potential issues
+
+Homeserver performance and storage may be impacted by this endpoint. While a homeserver already stores
+membership information for each of its users, the information may not be stored in a way that is readily
+accessible. Homeservers that have implemented [POST /user_directory/search](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3user_directorysearch)
+may have started some of this work, if they are limiting users to searching for users for which they
+share rooms. While this is not a given by any means, it may mean that implementations of this API
+and /search may be complimentary.
```
something we've been moving away from is pulling large chunks of data out of the server - a question to consider is if this endpoint needs pagination, or if it's unlikely to be a problem. 

Server-side folks probably have a stronger opinion on this than I.

ShadowJonathan replied:
I feel like this is not particularly "large", it's going to be a subset of rooms the user is in, and only then, just their IDs, which would be a string that's reasonably usually around 16 characters.

Even if the user would share around 100 rooms with the other person (a result on the high end, in my opinion), with 16 characters average, it'd only be around ~1.8kb to share (taking into account quotes and commas), which would still be less than a random average (synapse) sync response of around 2KB, and only double of its empty sync response (of ~700B).

Even if the numbers start to rise (beyond 1000 shared rooms), I don't think this'll ever get into the Megabyte territory of responses, and thus require pagination.

Thoughts?

erikjohnston replied:
I'd quite like us to move to a world where every endpoint that responds with a variable amount of data is paginated, for a few reasons:

1. it's much less faff for clients if the endpoints start off being paginated compared with introducing pagination later;
2. if everything is paginated and follow a similar structure then its just easier for clients and servers to deal with; and
3. even if the common case only returns a small amount of data, the mere potential that a client *can* cause lots of data to be returned makes it hard to run Synapse in a resource constrained way.

Also, if we end up extending the API to include stripped state then the size of the response can easily balloon.

ShadowJonathan replied:
/res this has been added in https://github.com/matrix-org/matrix-spec-proposals/pull/3913

We didn't need to specify the order the rooms are returned in, but allow the server to specify the ordering itself, and simply return an opaque token to get the next batch of rooms

this'd allow it to define an ordering itself, with any kind of optimisation strategy

turt2live commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,79 @@
+```json
+{
+  "joined": [
+    "!OGEhHVWSdvArJzumhm:matrix.org",
````
Should we be returning stripped state for these rooms too? (or do we plan to do something similar in a future release? if so, making this an object with an empty object as the value would give extensibility).

ShadowJonathan replied:
The idea of this endpoint is that it just gives pointers to rooms for an existing "spun-up" client, one which already would have information about these rooms.

However, I don't know the full details of sync v3, and how it'll handle this, will it effectively hide rooms "under the fold", including any and all detail?

If so, then I personally think there's indeed merit in making this future-proof (and thus making it a mapping to empty objects).

However still, if this is going to be the case, I have to think about pagination, which complicates this endpoint a lot. (See your previous comment and my response)

turt2live reviewed (commented):
drive-by review. Sorry, it's incomplete but should be enough to cover some concerns.

erikjohnston commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,79 @@
+
+Homeserver performance and storage may be impacted by this endpoint. While a homeserver already stores
+membership information for each of its users, the information may not be stored in a way that is readily
+accessible. Homeservers that have implemented [POST /user_directory/search](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3user_directorysearch)
```
I think the performance impact will be OK if this endpoint is used "infrequently", i.e. like for the use cases described in the intro.

Basically: the query requires looking up the list of rooms for each user and doing the intersection, which is not cheap, but is also not particularly expensive.

ShadowJonathan replied:
Yeah, in Element's case, I expect this endpoint to be hit in a few cases;
- Explicitly opening a detail profile view (e.g. clicking on a profile in the sidebar in cinny in element).
- Upon viewing invites (to cross-reference where the person is known from).
- When viewing a minimal DM, where the header ("This is your conversation with X") is visible.

I don't expect it to be pre-loaded en-masse, or loaded en-masse in general.

ShadowJonathan committed:
Apply review feedback (#3913)

- convert to using query parameters
- add pagination
- add rate-limit-ness
- add forward-compatibility considerations
- alter unstable prefix and unstable_features
- misc reflow of the markdown

ShadowJonathan commented:
Checkpoint for myself, and as a general note of progress:

With the changes introduced by #3913, the JS SDK and Synapse implementation are no longer valid, and need to be updated.

The following has to be implemented:
- The change from `/mutual_rooms/{user_id}` to `/mutual_rooms?user_id={user_id}`
- The addition of `next_batch_token` (and logic)
- Unstable flag now being `uk.half-shot.msc2666.query_mutual_rooms` (instead of `<...>.mutual_rooms`)

---

- [ ] Update the JS SDK to reflect this: #TBD
- [ ] Update Synapse to reflect this: #TBD
- [ ] Update Ruma to reflect this: #TBD

turt2live commented:
@ShadowJonathan given the status you posted, I'm removing this from the SCT's board for now. When it's (likely) ready for FCP again, please raise it in the [SCT Office](https://matrix.to/#/#sct-office:matrix.org) and we'll add it back.

clokep commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,115 @@
+```
+
+The server may decide that the response to this endpoint is too large, and thus an optional key
+`"next_batch_token"` can be inserted, which the client has to pass to `batch_token` in the query
````
The synapse implementation seems to return a particular error for unknown batch tokens?

ShadowJonathan replied:
...and merged, this thread can be resolved.

ShadowJonathan committed:
Add invalid batch_token error code (#4017)

ShadowJonathan commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,118 @@
+`uk.half-shot.msc2666.query_mutual_rooms`.
+
+Previous iterations of this MSC has used the following `unstable_features` key(s):
+- `uk.half-shot.msc2666.mutual_rooms`
```
🐑 Small addition, as I forgot this a few times prior:

```suggestion
- `uk.half-shot.msc2666.mutual_rooms`
- `uk.half-shot.msc2666`
```

anoadragon453 committed:
Update proposals/2666-get-rooms-in-common.md

Co-authored-by: Jonathan de Jong <jonathandejong02@gmail.com>

ShadowJonathan commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+  "joined": [
+    // ...
+  ],
+  "next_batch_token": "<any valid ascii string up to 32 chars>"
```
I'm just now realising that this might cause problems, and we should restrict this to just `[a-zA-Z0-9]` to prevent issues with url encoding.

No implementation (right now) generates a batch token, so if the below suggestion is accepted, no implementations would have to be altered.

```suggestion
  "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9]+/>"
```

kevincox replied:
It may be convenient to also include `-_` so that data can be URL-safe base64 encoded. This provides a bit of an escape hatch for encoding arbitrary data. This way 24 binary bytes can be encoded into the token. The token should of course be considered opaque to the client still.

anoadragon453 commented:
In my opinion, it would be nice to see an implementation of this MSC in a client before acceptance, just to ensure that the proposed spec is enough for a client to implement a decent UI. Such an implementation can simply exist on a fork of a codebase.

anoadragon453 committed:
restrict the allowed characters for the next_batch_token

Co-authored-by: Jonathan de Jong <jonathandejong02@gmail.com>

tulir commented:
The original Element Web implementation still works fine, I have it in my fork here: https://github.com/maunium/matrix-react-sdk/tree/tulir/shared-rooms

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+  "joined": [
+    // ...
+  ],
+  "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9]+/>"
```
Other spots in the spec call a similar field `next_batch` instead of `next_batch_token`. It seems confusing to deviate?

* https://spec.matrix.org/v1.7/client-server-api/#post_matrixclientv3search
* https://spec.matrix.org/v1.7/client-server-api/#pagination
* https://spec.matrix.org/v1.7/client-server-api/#get_matrixclientv1roomsroomidthreads
* etc

ShadowJonathan replied:
Apologies, I wasn't aware of the naming convention.

According to this, its both `next_batch` in the response, and `next_batch` in the query request.

I could update, but that would be a bunch more churn. If you like to see me do it, just say it.

anoadragon453 replied:
Implementations will need to update to remove unstable prefixes anyways, so why not do a parameter rename at the same time?

The implementations don't need to have the exact names to be considered proven and unblock FCP.

turt2live replied:
Pagination is also specified as https://spec.matrix.org/v1.7/appendices/#pagination ftr

ShadowJonathan replied:
/res #4035

Mikaela commented:
This is also implemented in Nheko unless there are multiple competing ways of displaying shared rooms.

ShadowJonathan commented:
@Mikaela There are, i recommend taking a look at the [latest JS SDK](https://github.com/matrix-org/matrix-js-sdk/pull/3381) commit to see how it handles the MSC's history, and how it approaches calling the unstable endpoint.

When it lands as stable, this could be removed and only the latest endpoint be required, but since i wanted to not remove perfectly-working backwards-compatible MSC changes (only cosmetic ones, like changing a path to a query, or changing the endpoint name itself), i implemented it anyways.

ShadowJonathan commented:
<details>
<summary>A general comment about my feelings for the usefulness of this MSC, and how it developed</summary>

When I began shepherding this MSC, I was anxious that it would be eventually rejected because sync would provide all the necessary details for a user to know which rooms a user would be in, but after this MSC aged (like fine cheese), and other developments came to light, or i realised other things, i realise that this endpoint does have merit.

Right now, in today's sync, we have lazy-loading members, which would already remove the guarantee that the client knows all users at any time. However, with sliding sync (if i'm not mistaken), we will also narrow that viewpoint even more, where the client doesnt even need to know about the existence of "older" rooms, like this.

As such, having an endpoint that definitively grabs a user's mutual rooms, would still hold up, and be useful in an age where matrix clients can opt into even-more-sparse synchronisation, to keep this as a feature that helps along privacy security, and helps a user judge abuse when they're invited out of the blue, and many more use cases.

</details>

ShadowJonathan commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+  "joined": [
+    // ...
+  ],
+  "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9]+/>"
```
Per [this comment](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r1218010122)

```suggestion
  "next_batch_token": "<a string up to 32 characters abiding by the regex /[a-zA-Z0-9_-]+/>"
```

turt2live replied:
can we just use https://github.com/matrix-org/matrix-spec-proposals/blob/rav/proposals/id_grammar/proposals/1597-id-grammar.md#opaque-ids please?

Ideally by name: "a batch_token is an opaque identifier, currently capable of containing the characters [0-9a-zA-Z._~-], must be at most 255 characters, and non-empty"

ShadowJonathan replied:
Thanks, I was looking for something like that, will add it 👍

ShadowJonathan replied:
/res #4035

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+
+Homeservers will implement a new endpoint `/_matrix/client/v1/user/mutual_rooms`.
+
+This endpoint will take a query parameter of `user_id` which will contain the MXID of the user
+matched against.
```
If `user_id` can't be specified multiple times, it might belong in the HTTP path instead. Also, can it be specified multiple times?

turt2live replied:
ah, it's listed later in the MSC. Is there a technical reason preventing multiple users from being searched? It feels like an awkward thing to make a future MSC for when we're already here.

ShadowJonathan replied:
The discussion context is here: https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r936483418

ShadowJonathan replied:
There is no technical reason preventing multiple users for being searched, but I don't see a convincing client usecase, while I agreed to keep the path open for one if the time came, to make it easily extendible.

richvdh replied:
I think if we're *not* putting the user id in the path, then we shouldn't use the `/_matrix/client/v1/user` hierarchy, since most of the endpoints under that hierarchy do have a user_id as the next path component.

ShadowJonathan replied:
I'm also realising that if we want to be able to get rooms in common with more users, there could just be another version of this endpoint.

KitsuneRal replied:
With the current MSC text, this seems to be resolved?

ShadowJonathan replied:
I want some clear direction on this. The current version (that I merge in this branch via #4035) changes some labels around enough that i'd have to give it another unstable_features tag.

I haven't yet, as i want to sweep the result of this in that tag, before i propose it FCP again.

---

Should I change this back to a path component, or should I keep this as a query element endpoint? I can do either.

Half-Shot replied:
This seems to be the stalling factor in this MSC now, and it seems like a fairly trivial matter of just:

 - Drop `/user/`, path becomes `_matrix/client/v1/mutual_rooms`
 - Keep the query parameter, leave the door open to allowing multiple in the future if the need arises.

If everyone is happy with that, we can make the change and re-request FCP.

tulir replied:
`/v1/mutual_rooms` rather than /v3, but yeah that sounds good to me. No need to define a new unstable prefix either, just change the stable one

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+This endpoint will take a query parameter of `user_id` which will contain the MXID of the user
+matched against.
+
+This endpoint can be rate limited.
```
boilerplate:
```suggestion
This endpoint can be rate limited and requires authentication.
```

ShadowJonathan replied:
/res #4035

turt2live commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+The server may decide that the response to this endpoint is too large, and thus an optional key
+`"next_batch_token"` can be inserted, which the client has to pass to `batch_token` in the query
+parameters together with the original `user_id` to fetch the next batch of responses. This will
+continue until the server does no longer insert `"next_batch_token"`.
````
Somewhere we need the words "this batch_token is only valid for use with this endpoint", and the semantics for expiration (does the homeserver need to remember the batch token forever?)

ShadowJonathan replied:
I think an expiration time of 10 minutes could be reasonable here.

anoadragon453 replied:
In which case we should specify an error case for clients supplying an unknown / expired batch token, so that they can retry.

We can't use `M_UNKNOWN_TOKEN` - as this error would also be returned if the user used an unknown access/refresh token to access this endpoint.

We may need to define a new `errcode` - perhaps `M_UNKNOWN_PAGINATION_TOKEN`?

ShadowJonathan replied:
#4035

I have added the 10 minute timer, and added text that it will handle expired tokens the same as invalid tokens, so that implementations may forget those tokens entirely. Do you think this is a good method?

Then the clients' default response to "invalid token" is just to start over, without any token.

The only problem i see is that clients may loop on a broken implementation, where it is not passing the right token back to the server will cause it to get stuck in its mechanism where it continually retries.

anoadragon453 replied:
I would just leave the validity period up to the implementation, but *recommend* at least 10 minutes. Clients should be prepared to handle tokens that do have a ridiculously short window though.

I would also recommend clients limit the maximum number of retries before giving up, precisely to prevent infinite loops as you describe.

turt2live commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+
+If the value is false or the key is not present, clients MUST assume the feature is not available.
+
+Once the MSC has been merged, and the homeserver has advertised support for the Matrix version that
+this endpoint is included in, clients should use `/_matrix/client/v1/user/mutual_rooms` and will no
+longer need to check for the `unstable_features` flag.
```
```suggestion
Once this MSC is considered stable, implementations can use the stable endpoint instead.
```

It may also be worth reading https://github.com/matrix-org/matrix-spec-proposals/pull/4024 to match the language there, as this seems like the sort of feature a client would want to use between FCP completing and spec release.

ShadowJonathan replied:
Ah, I think with this original language I remember/wrote this down to "consider stable" to be "when it is included in a spec version"

Together with richvdh's comment [here](https://github.com/matrix-org/matrix-spec-proposals/pull/4024#discussion_r1219377897), i think that for this MSC it is unnecessary to introduce a `.stable` `unstable_features` flag.

ShadowJonathan replied:
/res

anoadragon453 replied:
@tulir I assume you (and other clients) will want to use this before homeserver implementations get around to advertising support for the Matrix version this will merge into?

tulir replied:
Oh yes, stable flags are needed for basically all endpoints

turt2live reviewed (approved):
overall, the MSC seems like a sensible way forward. A couple FCP-blocking concerns:

* Stuff to do with pagination
* Text/discussion around supporting multiple `user_id` params

turt2live commented:
@mscbot fcp merge

turt2live commented:
@mscbot concern Pagination should match appendices
@mscbot concern Discussion around searching for multiple user IDs, and text to formalize the conclusion there

richvdh commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,119 @@
+rooms with the authenticated user, an empty array should be returned.
+
+```http
+GET /_matrix/client/v1/user/mutual_rooms/?user_id=%40bob%3Aexample.com
````
this example includes a trailing `/`, which is inconsistent with the definition above.

In general Matrix does not allow extra `/`s in endpoints.

ShadowJonathan replied:
Ah, that's an error from a previous change that moved path-element parameters to query parameters.

ShadowJonathan replied:
/res https://github.com/matrix-org/matrix-spec-proposals/pull/4035

richvdh commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+This endpoint can be rate limited.
+
+The response format will be an array containing all rooms where both the authenticated user and
+`user_id` have a membership of type `join`. If the `user_id` does not exist, or does not share any
+rooms with the authenticated user, an empty array should be returned.
```
is the behaviour on invalid user IDs left as an implementation decision? (I think that's fine if so, but best to say so explicitly)

ShadowJonathan replied:
IIRC that would fall under some `400` `Bad Request` semantics, implementation-specific, yeah.

richvdh commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,119 @@
+The response error for when the given `batch_token` is invalid will be a response with HTTP code 400,
+with `M_INVALID_PARAM` as `errcode`.
+
+The response error for trying to get shared rooms with yourself will be an HTTP code 422, with
+`M_INVALID_PARAM` as the `errcode`.
```
Not a strong opinion here, but 422 is pretty obscure (and we don't use it anywhere else in matrix, afaik).

The *idea* in Matrix is that we shouldn't rely too heavily on HTTP response codes, to make it easier (one day) to move to alternative transports. So if we need to distinguish this from other error cases, we'd do that via a custom `errcode` rather than HTTP response code.

KitsuneRal replied:
Ftr, [CoAP](https://www.rfc-editor.org/rfc/rfc7252#section-5.9) doesn't have code 4.22 or anything semantically similar to HTTP's 422, and there don't seem to be subsequent RFCs addressing that (while 429, e.g., has got is CoAP equivalent in a [later RFC](https://datatracker.ietf.org/doc/html/rfc8516)).

anoadragon453 replied:
Doing this sounds like programmer error vs. something that would actually need to be explicitly handled in practice. So this could just be a `400` / `M_UNKNOWN` - "you cannot request rooms in common with yourself".

ShadowJonathan replied:
I've changed the error to 400 + M_UNKNOWN

/res #4035

richvdh reviewed (commented):
Generally seems good to me but there are a few things I'd like to see resolved before FCP

turt2live commented:
It appears as though the author(s) aren't currently able to take this on - cancelling FCP pending the MSC being [re-raised to the SCT](https://matrix.to/#/#sct-office:matrix.org)

@mscbot fcp cancel

ShadowJonathan commented:
<details>
<summary>Current Status</summary>

With #4035, I have applied a lot of changes that will need a new unstable_features tag. However, I want [this discussion](https://github.com/matrix-org/matrix-spec-proposals/pull/2666#discussion_r1262342461) to be resolved first before I go ahead and do that.

After that, implementations need to be updated to match, and only then will I request FCP again.

</details>

ShadowJonathan committed:
apply review feedback (#4035)

Half-Shot commented:
Hello 5 years later :wave: 

https://github.com/element-hq/element-web/pull/30654 is an implementation that finally uses this information, so could be used as an example of an implementation.

Half-Shot committed:
Update path

tulir committed:
Link to opaque identifier spec for next_batch

tulir committed:
Revert unstable prefix path

tulir committed:
Define guest access

tulir commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,131 @@
+`"next_batch"` can be inserted, which the client has to pass to `from` in the query
+parameters together with the original `user_id` to fetch the next batch of responses. This will
+continue until the server does no longer insert `"next_batch"`. The `next_batch` field follows
+the [standard for opaque identifiers](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).
```
Actually, if the reason for adding pagination is primarily response size, then I can think of a relatively trivial implementation:

1. compute entire response list
2. sort list lexicographically
3. cut off response at a certain index
4. return the last room id as the next batch token
5. when a request has a from token, drop items that are lexicographically lower than the token before cutting the response list

tulir replied:
Only problem with that is the character set for opaque identifiers doesn't allow `!`, plus pre-v12 room IDs can have arbitrary unicode. Maybe it can be base64 encoded and truncated? Just need handling for the silly user case of having more than the batch size of pre-v12 rooms that have the same first 191 letters (e.g. don't return next batch token if new token matches old one)

tulir commented:
This has been mostly ready for over 2 years and with https://github.com/element-hq/synapse/pull/19279, everything should be implemented as well

@mscbot fcp merge

tulir committed:
Clarify handling invalid and non-existent user IDs

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,136 @@
+The server may decide that the response to this endpoint is too large, and thus an optional key
+`"next_batch"` can be inserted, which the client has to pass to `from` in the query
+parameters together with the original `user_id` to fetch the next batch of responses. This will
+continue until the server does no longer insert `"next_batch"`. The `next_batch` field follows
```
```suggestion
continue until the server no longer inserts `"next_batch"`. The `next_batch` field follows
```

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,136 @@
+The response error for trying to get shared rooms with yourself will be an HTTP code 400, with
+`M_UNKNOWN` as the `errcode`. And the error description may be "you cannot request rooms in common with yourself".
+
+Tokens generated by this endpoint must be valid for at least 10 minutes, after which, tokens can expire.
```
```suggestion
Batch tokens generated by this endpoint SHOULD be valid for at least 10 minutes, after which, tokens can expire.
```

It's impossible to enforce this in [the Synapse implementation](https://github.com/element-hq/synapse/pull/19279), which just uses the base64url-encoded, last-returned Room ID as the batch token. It will be invalidated by the user joining/leaving rooms or rooms being purged by the homeserver.

tulir committed:
Apply suggestions from code review

Co-authored-by: Andrew Morgan <1342360+anoadragon453@users.noreply.github.com>

tulir committed:
Add stable feature flag

tulir committed:
Allow omitting rooms from response when using batch tokens

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
````diff
@@ -0,0 +1,140 @@
+  "next_batch": "<an opaque identifier, containing only the characters [0-9a-zA-Z._~-], non-empty if not omitted, and at most 255 characters>"
+}
+```
+
````
```suggestion

The client can make another request to get the next batch of results:

```http
GET /_matrix/client/v1/user/mutual_rooms?user_id=%40bob%3Aexample.com&from=<opaque_identifier>
```
```

(Note: GitHub won't render the above properly due to the additional set of backticks.)

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,140 @@
+The batch tokens this endpoint generates are only valid for this endpoint.
+
+The response error for when the given `from` batch token is invalid will be a response with HTTP code 400,
+with `M_INVALID_PARAM` as `errcode`.
```
```suggestion
If the batch token provided by the client is deemed invalid by the server, the server
MAY return the standard error response with HTTP code 400 and `errcode`
`M_INVALID_PARAM`.
```

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,140 @@
+
+## Forward-compatibility considerations
+
+There possibly comes a time where it's desirable to query mutual rooms for more-than-one other user,
```
```suggestion
There possibly comes a time where it's desirable to query mutual rooms for more than one other user,
```

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,140 @@
+
+Because of that, the endpoint accepts a query parameter, however, it will only accept *one* query
+parameter for the time being. In the future this restriction can be lifted to accept multiple query
+parameters under `user_id`
```
```suggestion
parameter for the time being. This restriction may be lifted by a future MSC.
```

anoadragon453 committed:
Tighten up wording of the MSC

Make the language less fluffy and easier to read.

anoadragon453 commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,149 @@
+
+Homeservers will implement a new endpoint `/_matrix/client/v1/mutual_rooms`.
+
+This endpoint will take a query parameter of `user_id` which will contain the MXID of the user
```
```suggestion
This endpoint will take a required query parameter of `user_id` which will contain the MXID of the user
```

anoadragon453 reviewed (approved):
This MSC and the associated Synapse implementation look good to me.

Excited to see this one land!

anoadragon453 committed:
remove space in 'MSC2666'

clokep commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,149 @@
+
+Handling invalid user IDs may result in an error, is likely implementation-specific,
+and is beyond the scope of this proposal. Implementations are also free to simply
+return a success response with an empty array for invalid user IDs.
```
This paragraph is a bit weird. I'd suggest making it an error instead:

```suggestion
If an invalid user ID is provided the server MUST respond with a 400 error with
code `M_INVALID_PARAM`.
```

tulir replied:
Currently the MSC says to leave it as an implementation detail rather than encoding anything into the spec. Other endpoints like `GET /profile/{userId}` also don't specify any particular handling of invalid user IDs. I'd prefer doing it consistently for all endpoints rather than just this one.

uhoreg replied:
@mscbot concern should an error be required for invalid user IDs

clokep replied:
Then let's just spec something sane as an edge case instead of leaving it up to the implementation.

tulir commented on proposals/2666-get-rooms-in-common.md:
```diff
@@ -0,0 +1,149 @@
+The server may decide that the response to this endpoint is too large, and only return
+a subset of the results. In this case, the server should populate the optional field
+`"next_batch"` with an
+[opaque identifier](https://spec.matrix.org/v1.16/appendices/#opaque-identifiers).
```
In the current implementations adding the count would be trivial. It might not be in future implementations, but maybe that doesn't matter? The primary goal of pagination is ensuring the server doesn't have to send arbitrarily large response payloads.

We could define that the count may be inaccurate or something like that to let servers skip calculating the exact count if they feel it's too difficult

tulir committed:
Apply suggestions from code review

Co-authored-by: Andrew Morgan <1342360+anoadragon453@users.noreply.github.com>
Co-authored-by: David Baker <dbkr@users.noreply.github.com>

tulir committed:
Add count field
